### ISSUE TITLE
Support WP-Stash in 'wp cache type'

### ISSUE THREAD

**Issue by Biont** [wp-cli/cache-command#68]
Greetings.

I just looked around the `wp cache type` command after seeing it returns "Unknown" in our WP instances running [WP Stash](https://github.com/inpsyde/WP-Stash). I think it would be pretty useful if we could see that a site is indeed running WP-Stash - and possibly which driver is in use.

I see that determining the cache type [is a "manual" process](https://github.com/wp-cli/wp-cli/blob/70e0be2f238ac59b4794cc3ef040ca0abdac268d/php/utils-wp.php#L171-L224) that checks for the specifics of implementations. So would you be supportive of adding WP Stash to the list of implementations to check against?

The most basic check would probably be `class_exists('Inpsyde\WpStash\WpStash')` and it could return `get_class( WpStash::instance()->driver() )`. This would yield the FQCN of the [StashPHP driver](https://www.stashphp.com/Drivers.html) that is currently being used.

I hope you consider this a worthwhile addition - which I'd happily PR your way.

As a side note: Wouldn't it make sense to provide a generic way for implementations to TELL wp-cli what's being used? The ideal solution would be of course if WP-Core had such a thing on offer, but in the absence of this: is there an inherent problem with checking for the existence of `$wp_object_cache->type` - or maybe even calling `wp_cache_type()` if it exists?

In addition, maybe you could return `'Unknown: ' . get_class( $wp_object_cache )`, so users with entirely custom solutions would at least get *some* result.

**Comment by danielbachhuber** [comment]
Thanks for the suggestion, @Biont !

I'd be happy to accept a PR for what you're suggesting.

> As a side note: Wouldn't it make sense to provide a generic way for implementations to TELL wp-cli what's being used? The ideal solution would be of course if WP-Core had such a thing on offer, but in the absence of this: is there an inherent problem with checking for the existence of `$wp_object_cache->type` - or maybe even calling `wp_cache_type()` if it exists?

If a `wp_cache_type()` function existed in WordPress core, we could easily call it. I think it makes most sense to implement that formally first. It's not too much effort to maintain a list of object caches as there aren't many of them.

> In addition, maybe you could return `'Unknown: ' . get_class( $wp_object_cache )`, so users with entirely custom solutions would at least get _some_ result.

Sure! Feel free to submit that as a second PR if you'd like.

**Comment by danielbachhuber** [comment]
Related https://github.com/wp-cli/wp-cli/issues/5735

**Comment by OllieJones** [comment]
`wp cache type` is broken with the SQLite Object Cache too.
I propose the following solution.

1. New and revised persistent object cache plugins should provide a member function like this one. https://github.com/OllieJones/sqlite-object-cache/blob/0be3dbf6d6e9c7d3fb958964e1d169db191eeae0/assets/drop-in/object-cache.php#L1961 

```
		/**
		 * Return the cache type. For use by "wp-cli cache type" and other display code.
		 *
		 * @return string The type of cache, "SQLite".
		 */
		public function get_cache_type() {
			return 'SQLite';
		}
```

2. WP-CLI should do the if (defined) call pattern on `$wp_object_cache->get_cache_type()` to get its results...
3. Failing that, use the existing method.

This will prove more extensible that the current method.

**Comment by danielbachhuber** [comment]
> 2. WP-CLI should do the if (defined) call pattern on `$wp_object_cache->get_cache_type()` to get its results...

I think this needs to exist in core first. I don't want to introduce support for something official-looking, only to have core implement some different approach.

**Comment by swissspidy** [comment]
Opened https://core.trac.wordpress.org/ticket/58808 for that proposal

**Comment by tillkruss** [comment]
@swissspidy I worked a bit with @johnbillion on a spec for object cache information, in form of `wp_cache_info()`. Do you want to collaborate on this?

**Comment by swissspidy** [comment]
Yeah https://github.com/rhubarbgroup/wp-object-cache-info-spec was just mentioned to me today. I think I'm the wrong one to collaborate on that though, as I am not really familiar with all the object cache stuff.
